src/testing/systest/supervisor/logged_process.zig

```diff
@@ -0,0 +1,299 @@
+    self.child.stdout_behavior = .Ignore;
+    self.child.stderr_behavior = .Inherit;
+
+    try self.child.spawn();
+
+    errdefer {
+        _ = self.child.kill() catch {};
+    }
+
+    // Zig doesn't have non-blocking version of child.wait, so we use `BrokenPipe`
+    // on writing to child's stdin to detect if a child is dead in a non-blocking
+    // manner. Checks once a second in a separate thread.
+    _ = try std.posix.fcntl(
+        self.child.stdin.?.handle,
+        std.posix.F.SETFL,
+        @as(u32, @bitCast(std.posix.O{ .NONBLOCK = true })),
+    );
+    self.stdin_thread = try std.Thread.spawn(
+        .{},
+        struct {
+            fn poll_broken_pipe(stdin: std.fs.File, process: *Self) void {
+                while (process.state() == .running) {
+                    std.time.sleep(1 * std.time.ns_per_s);
+                    _ = stdin.write(&.{1}) catch |err| {
+                        switch (err) {
+                            error.WouldBlock => {}, // still running
+                            error.BrokenPipe,
+                            error.NotOpenForWriting,
+                            => {
+                                // Only write the state variable in case the process is still
+                                // considered running. If it was actively terminated, we don't
+                                // want to overwrite that state.
+                                _ = process.current_state.cmpxchgStrong(
+                                    .running,
+                                    .completed,
+                                    .seq_cst,
+                                    .seq_cst,
+                                );
+                                break;
+                            },
+                            else => @panic(@errorName(err)),
+                        }
+                    };
+                }
+            }
+        }.poll_broken_pipe,
+        .{ self.child.stdin.?, self },
+    );
+
+    return self;
+}
+
+pub fn destroy(self: *Self) void {
+    assert(self.state() == .terminated or self.state() == .completed);
+    const allocator = self.allocator;
+    allocator.free(self.cwd);
+    allocator.destroy(self);
+}
+
+pub fn state(self: *Self) State {
+    return self.current_state.load(.seq_cst);
+}
+
+pub fn terminate(
+    self: *Self,
+) !std.process.Child.Term {
+    self.expect_state_in(.{.running});
+    defer self.expect_state_in(.{.terminated});
+
+    // Terminate the process.
+    //
+    // Uses the same method as `src/testing/tmp_tigerbeetle.zig`.
+    // See: https://github.com/ziglang/zig/issues/16820
+    _ = kill: {
+        if (builtin.os.tag == .windows) {
+            const exit_code = 1;
+            break :kill std.os.windows.TerminateProcess(self.child.id, exit_code);
+        } else {
+            break :kill std.posix.kill(self.child.id, std.posix.SIG.KILL);
+        }
+    } catch |err| {
+        std.debug.print(
```

Comment:
log.err

In general, never use `std.debug.print` (I used it a bunch in various auxiliary support files, and I now regret that 🤣 )